lnwallet/channel.go

```diff
@@ -6028,3 +6038,107 @@ func (lc *LightningChannel) RemoteCommitHeight() uint64 {
 func (lc *LightningChannel) FwdMinHtlc() lnwire.MilliSatoshi {
 	return lc.channelState.LocalChanCfg.MinHTLC
 }
+
+// calcIncomingHtlcs calculates the number of committed and uncommitted HTLC's
+// from the PoV of the tip of the commitment chain. This function is used to
+// determine if one of the counterparty's adds will overflow the commitment
+// transaction before it happens in ReceiveNewCommitment. Since these Adds will
+// be appended to the tip of the chain, the tail is not used. It takes an
+// htlcID parameter that is used to determine the number of uncommitted HTLC's
+// in-memory.
+func (lc *LightningChannel) calcIncomingHtlcs(htlcID uint64) uint64 {
+	// There are two cases to consider when calculating the true number of
+	// committed htlc's. In both cases, it is assumed that Bob has the maximum
+	// number of incoming HTLC's at the start.
+	//
+	// CASE 1:
+	//
+	// 	Alice                    Bob
+	// 1.         <---settle/fail---
+	// 2.         <-------sig-------
+	// 3.         --------rev------>
+	// 4.         --------sig------>
+	// 5.         <-------rev-------
+	// 6.         --------add------>
+	//
+	// Bob should successfully receive the last add in step 6 since a
+	// signature was received in step 4 which removes a prior add via the
+	// settle / fail in step 1.
+	//
+	//
+	// CASE 2:
+	//
+	// 	Alice                    Bob
+	// 1.         <---settle/fail---
+	// 2.         <-------sig-------
+	// 3.	      --------sig------> (covers an add sent before step 1)
+	// 4. 	      <-------rev-------
+	// 5.         --------rev------>
+	// 6.         --------add------>
+	// 7.         - - - - sig - - ->
+	//
+	// This case represents a more asynchronous commitment dance in which both
+	// sides are sending signatures at the same time. In step 3, the signature
+	// does not cover the recent settle/fail that Bob sent in step 1. Because
+	// of this, the add Alice sends to Bob in step 6 could overflow Bob's
+	// commitment transaction depending on our calculation. If we just count
+	// the number of committed HTLC's incoming to Bob and the number of in-mem
+	// HTLC's, then the add in step 6 will overflow and Bob will fail the
+	// channel. If, however, we adjust the committed HTLC's count to ignore
+	// HTLC's which will be removed in the next signature that Alice sends,
+	// then the add in step 6 won't overflow. This is because the signature
+	// received in step 7 will cover the settle/fail in step 1 and makes space
```

Comment:
excellent comments, very helpful!

Reply:
I updated the code to just use `validateCommitmentSanity` since it was doing the same thing. I didn't feel the need to add a unit test for the specific case outlined above since it's not a new function like `calcIncomingHtlcs`. Curious what you think.